### PR TITLE
✨ Allow for an "All Time" view in statistics page

### PR DESCRIPTION
**Pull request by gBL17**

### CODE REVIEW

**Comment by gBL17** [comment]
https://github.com/user-attachments/assets/e784dd4a-59d3-4e1f-a1b5-09bc189a3381

File: app/src/main/kotlin/br/com/colman/petals/statistics/graph/AllTimeGraph.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import com.github.mikephil.charting.data.LineDataSet
+import java.time.LocalDate
+import java.time.YearMonth
+
+@Composable
+fun AllTimeGraph(uses: List<Use>, dateFormat: String) {
+  val description = stringResource(string.grams_distribution_over_days_since_first_use)
+  val gramsData = createAllTimeDistribution(uses)
+  val gramsDataList = mutableListOf<LineDataSet>()
+  gramsDataList.add(gramsData)
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/statistics/graph/AllTimeGraph.kt:20]:
```suggestion
  val gramsDataList = listOf(gramsData)
```

File: app/src/main/kotlin/br/com/colman/petals/statistics/graph/AllTimeGraph.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  LineChart(gramsDataList, description, 5f) {
+    axisMinimum = 1f
+    labelCount = 5
+    granularity = 1f
+    valueFormatter = DaysSinceFirstUseFormatter(uses, dateFormat).formatDate
+    addLimitLine(LimitLine(YearMonth.from(LocalDate.now()).monthValue.toFloat()).apply { lineWidth = 2f })
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/statistics/graph/AllTimeGraph.kt:27]:
```suggestion
    addLimitLine(LimitLine(YearMonth.now().monthValue.toFloat()).apply { lineWidth = 2f })
```

File: app/src/main/kotlin/br/com/colman/petals/statistics/graph/component/LineChart.kt [hunk 3]
```diff
@@ -20,6 +20,7 @@ import com.github.mikephil.charting.data.LineDataSet
 fun LineChart(
   datasets: List<LineDataSet>,
   description: String,
+  range: Float? = null,
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/statistics/graph/component/LineChart.kt:23]:
```suggestion
  visibleXRange: Float? = null,
```

File: app/src/main/kotlin/br/com/colman/petals/statistics/graph/component/LineChart.kt [hunk 4]
```diff
@@ -34,6 +35,13 @@ fun LineChart(
       chart.notifyDataSetChanged()
       chart.invalidate()
 
+      range?.let {
+        chart.setVisibleXRange(range, range)
+        chart.setVisibleXRangeMaximum(datasets[0].xMax)
+        chart.moveViewToX(datasets[0].xMax)
+        chart.notifyDataSetChanged()
+      }
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/statistics/graph/component/LineChart.kt:43]:
```suggestion
      visibleXRange?.let {
        chart.setVisibleXRange(it, it)
        chart.setVisibleXRangeMaximum(datasets[0].xMax)
        chart.moveViewToX(datasets[0].xMax)
        chart.notifyDataSetChanged()
      }
```

File: app/src/main/kotlin/br/com/colman/petals/statistics/graph/data/DistributionPerDaySinceFirstUse.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    setDrawValues(true)
+    fillColor = Green.toArgb()
+    color = Green.toArgb()
+    valueTextColor = White.toArgb()
+    valueTextSize = 14f
+  }
+}
+
+private fun calculateAllTimeGramsDistribution(uses: List<Use>): List<Entry> {
+  val dayBeforeFirstUseDate = uses.minBy { it.date }.localDate.toEpochDay()
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/statistics/graph/data/DistributionPerDaySinceFirstUse.kt:37]:
```suggestion
  val firstUseDate = uses.minBy { it.date }.localDate.toEpochDay()
```

File: app/src/main/kotlin/br/com/colman/petals/statistics/graph/data/DistributionPerDaySinceFirstUse.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  val daysSinceFirstUse = (dayBeforeFirstUseDate..now().toEpochDay()).toList()
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/statistics/graph/data/DistributionPerDaySinceFirstUse.kt:38]:
```suggestion
  val daysSinceFirstUse = (firstUseDate..now().toEpochDay()).toList()
```

File: app/src/main/kotlin/br/com/colman/petals/statistics/graph/formatter/DaysSinceFirstUseFormatter.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import com.github.mikephil.charting.formatter.IAxisValueFormatter
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+
+class DaysSinceFirstUseFormatter(uses: List<Use>, dateFormat: String) {
+
+  val formatDate = object : IAxisValueFormatter {
+    override fun getFormattedValue(value: Float, axis: AxisBase?): String {
+      val formatter = DateTimeFormatter.ofPattern(dateFormat)
+      val dayBeforeFirstUseDateToEpochDay: Long = uses.minBy { it.date }.localDate.toEpochDay().dec()
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/statistics/graph/formatter/DaysSinceFirstUseFormatter.kt:14]:
```suggestion
      val dayBeforeFirstUseDateToEpochDay: Long = uses.minBy { it.date }.localDate.toEpochDay() - 1
```

File: app/src/main/kotlin/br/com/colman/petals/statistics/graph/data/DistributionPerDaySinceFirstUse.kt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    valueTextSize = 14f
+  }
+}
+
+private fun calculateAllTimeGramsDistribution(uses: List<Use>): List<Entry> {
+  val firstUseDate = uses.minBy { it.date }.localDate.toEpochDay()
+  val daysSinceFirstUse = (firstUseDate..now().toEpochDay()).toList()
+  return daysSinceFirstUse.associateWith { uses.filter { u -> u.localDate.toEpochDay() == it } }.toSortedMap()
+    .map { (k, v) ->
+      Entry((k - dayBeforeFirstUseDate).toFloat(), v.totalGrams.toFloat())
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/statistics/graph/data/DistributionPerDaySinceFirstUse.kt:41]:
```suggestion
  val firstUseDay = uses.minBy { it.date }.localDate.toEpochDay()
  val daysSinceFirstUse = (firstUseDate..now().toEpochDay()).toList()
  return daysSinceFirstUse.associateWith { uses.filter { u -> u.localDate.toEpochDay() == it } }.toSortedMap()
    .map { (k, v) ->
      Entry((k - firstUseDay).toFloat(), v.totalGrams.toFloat())
```

**Pull request merged**